In "Overcoming confounding plate effects in differential expression analyses of single-cell RNA-seq data" \cite{Lun_2016} the same authors extend the negative-binomial model in different ways. In the Lun 2 simulation the cell factors are replaced with a library size factor and an additional level of variance is added by a batch effects factor. While the library size factor acts on individual cells the batch effects are applied to groups of cells from the same batch. Instead of simulating preliminary values, such as gene means, from statistical distributions the Lun 2 simulation samples them directly from a real dataset. Differential expression can be added between two sets of batches and the user can choose to use a zero-inflated model. If this is the case zero-inflated estimates of gene means and dispersions are used and an additional step randomly sets some counts to zero.
