In "Overcoming confounding plate effects in differential expression analyses of single-cell RNA-seq data" \cite{Lun_2016} the same authors extend the negative-binomial model from the Lun simulation. This simulation samples input parameters from real data, with very little random sampling from statistical distributions. In the Lun 2 simulation the cell factors are replaced with a library size factor and an additional level of variation is added by including a batch effects factor. While the library size factor acts on individual cells the batch effects are applied to groups of cells from the same batch. This simulation is thus highly specific to the scenario when there are known batch effects present in the data, for example Fluidigm C1 plate effects. Differential expression can be added between two sets of batches and the user can choose to use a zero-inflated negative-binomial (ZINB) model. Counts are simulated using the library size and plate factor adjusted gene means and the gene-wise dispersion estimates obtained from the data. If the ZINB model is chosen, zero inflated estimates of gene means and dispersions are used and an additional step randomly sets some counts to zero, based on the gene-wise proportions of zeroes observed in the data. Supplementary Figure 3 shows the model assumptions and parameters for this simulation.
