Splatter is designed to provide a consistent interface to multiple models for simulating scRNA-seq expression data. Each simulation consists of two steps. The first takes a real dataset and attempts to estimate the parameters required for the simulation and the second takes the estimated parameters, along with any additional parameters that cannot be estimated or are specified by the user, and generates a synthetic scRNA-seq dataset. The result of the first step is a parameters object unique to each simulation model. These objects have been designed to hold the required information in a sensible way and display to the user details such as which parameters can be estimated and which have been changed from the default value. It is important that each simulation have it's own object for storing parameters as they can vary greatly in the information they require, for example some simulations only need parameters for well known statistical distributions while others require large vectors or matrices of data sampled from real datasets. The results of the simulation step are returned as an SCESet object from the scater package \cite{McCarthy_2016}. The structure of this object is described in detail in the package documentation and the accompanying
paper but briefly it combines cell by feature matrices for storing expression values with tables for storing metadata about cells and features. This format makes it convenient to return intermediate values created during simulation as well as the final expression matrix and each of the simulations in Splatter does this. For example the underlying gene means in different groups are returned and could be used as a truth when evaluating differential expression testing. The documentation for each simulation describes the simulation model but also the parameters it uses and the intermediate values it returns.

An additional feature of Splatter is the ability to compare SCESet objects. These may be simulations with different parameters or models but could also be the real datasets that parameters were estimated from. The comparison function takes one or more SCESet objects, combines them (keeping any metadata that is present in all of them) and produces a series of plots that compare some of the aspects of scRNA-seq data. The combined datasets are also returned making it easy for the user to produce any additional comparison plots. 