---
title: "Simulations"
author: "Luke Zappia"
date: '`r Sys.Date()`'
output: html_document
---

Code version: `r system("git log -1 --format=oneline | cut -d' ' -f1", intern = TRUE)`

```{r knitr, include = FALSE}
DOCNAME = "simulations"
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = TRUE,
                      cache.path     = paste0("cache/", DOCNAME, "/"),
                      cache.comments = TRUE,
                      echo           = FALSE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("figures/", DOCNAME, "/"),
                      fig.width      = 10,
                      fig.height     = 8,
                      dev            = c('svg', 'png'),
                      message        = FALSE,
                      warning        = FALSE)
```

```{r libraries}
# RNA-seq
library("scater")
library("splatter")

# Plotting
library("mcriPalettes")

library("tidyverse")
```

```{r source, cache = FALSE}
source("../R/load.R")
source("../R/add_stats.R")
```

```{r ggtheme, cache = FALSE}
theme_doc <- ggplot2::theme_minimal(base_size = 18) +
             ggplot2::theme(panel.grid = ggplot2::element_blank(),
                            plot.title = ggplot2::element_text(size = 30))
ggplot2::theme_set(theme_doc)

my.pal <- mcriPalettes::mcriPalette("themes")
my.pal[4] <- "#F7DC5F"
```

Introduction
============

In this document we are going to compare some of the simulations available in 
Splatter to a real dataset and each other. The real data will come from the
study "Batch effects and the effective design of single-cell gene expression
studies" by Tung et al. Each simulation will estimate parameters from this
dataset. The majority of this document will be a series of plots comparing the
simulations.

Data
====

The Tung dataset contains data from three individuals, each of which have been
processed in three C1 plates. The reads have been aligned (using Subread) and
counted both as individual reads and using the UMIs. For this comparison we
are going to select a single individual (NA19239) and use the version of the
data that already been filtered to remove bad cells and lowly expressed genes.
We are also going to remove the ERCC spike-ins (because we won't simulate
those).

```{r real}
datasets <- list()
descrips <- list()

real.name <- "Tung"
real.desc <- "Single individual (filtered)"

real <- loadTung(count = "molecules", version = "filter",
                 individual = "NA19239", remove.ercc = TRUE)
cpm(real) <- edgeR::cpm(counts(real))
real <- addGeneStats(real, "counts")
real <- addGeneStats(real, "cpm")
real <- addGeneStats(real, "cpm", log = TRUE)
real <- addIsZeroOutlier(real)

datasets[[real.name]] <- real
descrips[[real.name]] <- real.desc
```

After selecting an individual and filtering we are left with `r nrow(real)`
genes and `r ncol(real)` cells in our real dataset.

Simulations
===========

```{r splatter}
sim.name <- "Splat"
sim.desc <- "Splatter simulation"

params <- splatEstimate(real)
sim <- splatSimulate(params, dropout.present = FALSE, seed = 1)
sim <- calculateQCMetrics(sim)
```

```{r add-splatter}
datasets[[sim.name]] <- sim
descrips[[sim.name]] <- sim.desc
```

```{r simple}
sim.name <- "Simple"
sim.desc <- "Simple negative-binomial"

params <- simpleEstimate(real)
sim <- simpleSimulate(params, seed = 1)
sim <- calculateQCMetrics(sim)
```

```{r add-simple}
datasets[[sim.name]] <- sim
descrips[[sim.name]] <- sim.desc
```

```{r lun}
sim.name <- "Lun"
sim.desc <- "Negative-binomial with cell factors"

params <- lunEstimate(real)
sim <- lunSimulate(params, seed = 1)
sim <- calculateQCMetrics(sim)
```

```{r add-lun}
datasets[[sim.name]] <- sim
descrips[[sim.name]] <- sim.desc
```

```{r lun2}
sim.name <- "Lun 2"
sim.desc <- "Negative-binomial with plate effects"

params <- lun2Estimate(real, plates = factor(pData(real)$replicate),
                       min.size = 50)
sim <- lun2Simulate(params, seed = 1)
sim <- calculateQCMetrics(sim)
```

```{r add-lun2}
datasets[[sim.name]] <- sim
descrips[[sim.name]] <- sim.desc
```

```{r scDD}
sim.name <- "scDD"
sim.desc <- "scDD simulation"

params <- scDDEstimate(real,
                       conditions = sample(1:2, ncol(real), replace = TRUE))
sim <- scDDSimulate(params, nCells = round(ncol(real) / 2), nDE = 0, nDP = 0,
                    nDM = 0, nDB = 0, nEE = nrow(real) / 2,
                    nEP = nrow(real) / 2, seed = 1)
sim <- calculateQCMetrics(sim)
```

```{r add-scDD}
datasets[[sim.name]] <- sim
descrips[[sim.name]] <- sim.desc
```

We are considering the following simulations:

```{r sim-details, cache = FALSE}
descrips <- unlist(descrips)
details  <- data.frame(Dataset     = names(descrips),
                       Type        = c("Real",
                                       rep("Simulation", length(descrips) - 1)),
                       Description = descrips)
knitr::kable(details, row.names = FALSE)
```

```{r combine}
fData.long <- mutate(fData(real), dataset = real.name)
pData.long <- mutate(pData(real), dataset = real.name)

for (sim.name in names(datasets)[-1]) {
    sim <- datasets[[sim.name]]
    cpm(sim) <- edgeR::cpm(counts(sim))
    sim <- addGeneStats(sim, "counts")
    sim <- addGeneStats(sim, "cpm")
    sim <- addGeneStats(sim, "cpm", log = TRUE)
    sim <- addIsZeroOutlier(sim)
    fData.long <- bind_rows(fData.long,
                            {fData(sim) %>% mutate(dataset = sim.name)})
    pData.long <- bind_rows(pData.long,
                            {pData(sim) %>% mutate(dataset = sim.name)})
}

fData.long <- mutate(fData.long,
                     dataset = factor(dataset, levels = names(datasets)))
pData.long <- mutate(pData.long,
                     dataset = factor(dataset, levels = names(datasets)))

saveRDS(fData.long, file = "../output/fData_long.Rds")
saveRDS(pData.long, file = "../output/pData_long.Rds")
```

Comparison
==========

Means
-----

```{r means}
ggplot(fData.long, aes(x = dataset, y = mean_log_cpm, colour = dataset)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75), size = 1.5) +
    scale_color_manual(values = my.pal) +
    ylab(expression(paste("Mean ", log[2], "(CPM + 1)"))) +
    ggtitle("Distribution of mean expression") +
    theme(legend.position = "none",
          axis.title.x = element_blank())

```

Variances
---------

```{r variances}
ggplot(fData.long, aes(x = dataset, y = var_cpm, colour = dataset)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75), size = 1.5) +
    scale_y_log10(labels = scales::comma) +
    scale_color_manual(values = my.pal) +
    ylab("CPM Variance") +
    ggtitle("Distribution of variance") +
    theme(legend.position = "none",
          axis.title.x = element_blank())
```

Mean-Variance
-------------

```{r mean-var}
ggplot(fData.long, aes(x = mean_log_cpm, y = var_log_cpm,
                       colour = dataset, fill = dataset)) +
    geom_point(size = 0.5, alpha = 0.1) +
    geom_smooth() +
    scale_color_manual(values = my.pal, name = "Dataset") +
    scale_fill_manual(values = my.pal, name = "Dataset") +
    xlab(expression(paste("Mean ", log[2], "(CPM + 1)"))) +
    ylab(expression(paste("Variance ", log[2], "(CPM + 1)"))) +
    ggtitle("Mean-variance relationship")
```

```{r mean-var-simple}
ggplot(fData.long, aes(x = mean_log_cpm, y = var_log_cpm,
                       colour = dataset, fill = dataset)) +
    geom_smooth() +
    scale_color_manual(values = my.pal, name = "Dataset") +
    scale_fill_manual(values = my.pal, name = "Dataset") +
    xlab(expression(paste("Mean ", log[2], "(CPM + 1)"))) +
    ylab(expression(paste("Variance ", log[2], "(CPM + 1)"))) +
    ggtitle("Mean-variance relationship") +
    theme(legend.position = "none")

```

```{r mean-var-simple-legend}
ggplot(fData.long, aes(x = mean_log_cpm, y = var_log_cpm,
                       colour = dataset, fill = dataset)) +
    geom_smooth() +
    scale_color_manual(values = my.pal, name = "Dataset") +
    scale_fill_manual(values = my.pal, name = "Dataset") +
    xlab(expression(paste("Mean ", log[2], "(CPM + 1)"))) +
    ylab(expression(paste("Variance ", log[2], "(CPM + 1)"))) +
    ggtitle("Mean-variance relationship") +
    theme(legend.position = "right")

```

Library size
------------

```{r lib-size}
ggplot(pData.long, aes(x = dataset, y = total_counts, colour = dataset)) +
    geom_boxplot(size = 1.5) +
    scale_y_continuous(labels = scales::comma) +
    scale_color_manual(values = my.pal) +
    ylab("Total counts per cell") +
    ggtitle("Distribution of library sizes") +
    theme(legend.position = "none",
          axis.title.x = element_blank())

```

Zeros
-----

```{r zeros-gene}
ggplot(fData.long, aes(x = dataset, y = pct_dropout, colour = dataset)) + 
    geom_boxplot(size = 1.5) +
    scale_color_manual(values = my.pal) +
    ylab("Percentage zeros per gene") +
    ggtitle("Distribution of zeros per gene") +
    theme(legend.position = "none",
          axis.title.x = element_blank())
```

```{r zeros-cell}
ggplot(pData.long, aes(x = dataset, y = pct_dropout, colour = dataset)) +
    geom_boxplot(size = 1.5) +
    scale_y_continuous(limits = c(0, 100)) +
    scale_color_manual(values = my.pal) +
    ylab("Percentage zeros per cell") +
    ggtitle("Distribution of zeros per cell") +
    theme(legend.position = "none",
          axis.title.x = element_blank())
```

Different situations
==================

Single
------

```{r single}
sim.single <- splatSimulateSingle(groupCells = 200, dropout.present = FALSE)
sim.single <- plotPCA(sim.single, return_SCESet = TRUE, draw_plot = FALSE)
```

```{r single-PCA}
data.frame(PC1 = reducedDimension(sim.single)[, 1],
           PC2 = reducedDimension(sim.single)[, 2]) %>%
    ggplot(aes(x = PC1, y = PC2)) +
    geom_point(size = 4, colour = "#092F5E")

```

Groups
------

```{r groups}
sim.groups <- splatSimulateGroups(groupCells = c(200, 100, 80, 20, 300, 120),
                                  de.prob = c(0.005, 0.005, 0.008, 0.005, 0.006, 0.005),
                                  dropout.present = FALSE)
sim.groups <- plotPCA(sim.groups, return_SCESet = TRUE, draw_plot = FALSE)
```

```{r groups-PCA}
data.frame(PC1 = reducedDimension(sim.groups)[, 1],
           PC2 = reducedDimension(sim.groups)[, 2],
           group = pData(sim.groups)$Group) %>%
    ggplot(aes(x = PC1, y = PC2, colour = group)) +
    geom_point(size = 4) +
    scale_color_manual(values = my.pal, name = "Group") +
    ggtitle("Simulation with groups")
```

```{r groups-PCA-notitle}
data.frame(PC1 = reducedDimension(sim.groups)[, 1],
           PC2 = reducedDimension(sim.groups)[, 2],
           group = pData(sim.groups)$Group) %>%
    ggplot(aes(x = PC1, y = PC2, colour = group)) +
    geom_point(size = 4) +
    scale_color_manual(values = my.pal, name = "Group")

```

Paths
-----

```{r paths}
sim.paths <- splatSimulatePaths(groupCells = c(100, 100, 100),
                                path.from = c(0, 1, 1),
                                dropout.present = FALSE, seed = 10)
sim.paths <- plotPCA(sim.paths, return_SCESet = TRUE, draw_plot = FALSE)
```

```{r paths-PCA}
data.frame(PC1 = reducedDimension(sim.paths)[, 1],
           PC2 = reducedDimension(sim.paths)[, 2],
           group = pData(sim.paths)$Group,
           step = pData(sim.paths)$Step) %>%
    ggplot(aes(x = PC1, y = PC2, colour = step, shape = group)) +
    geom_point(size = 4) +
    scale_colour_gradient(low = mcriPalette("themes")[2],
                          high = mcriPalette("themes")[1],
                          name = "Step") +
    scale_shape_manual(name = "Path", values = c(16, 17, 18)) +
    ggtitle("Simulation with paths")
```

```{r paths-PCA-notitle}
data.frame(PC1 = reducedDimension(sim.paths)[, 1],
           PC2 = reducedDimension(sim.paths)[, 2],
           group = pData(sim.paths)$Group,
           step = pData(sim.paths)$Step) %>%
    ggplot(aes(x = PC1, y = PC2, colour = step, shape = group)) +
    geom_point(size = 4) +
    scale_colour_gradient(low = mcriPalette("themes")[2],
                          high = mcriPalette("themes")[1],
                          name = "Step") +
    scale_shape_manual(name = "Path", values = c(16, 17, 18))
```

Session info
============

```{r session-info, cache = FALSE}
devtools::session_info()
```

```{r cleanup-docs, cache = FALSE}
doc.files <- c(list.files(pattern = "pdf"),
               list.files(pattern = "html"),
               list.files(pattern = "docx"))

for (file in doc.files) {
    file.rename(file, file.path("docs", file))
}
```

